tests/test_grid3d/test_grid_grdecl.py

```diff
@@ -0,0 +1,366 @@
+        with open_grdecl(mock_file, keywords=["SPECGRID"]) as kw:
+            keyword, values = next(kw)
+            assert keyword == "SPECGRID"
+            specgrid = ggrid.SpecGrid.from_grdecl(values)
+            assert specgrid.ndivix == 64
+            assert specgrid.ndiviy == 118
+            assert specgrid.ndiviz == 263
+            assert specgrid.numres == 1
+            assert specgrid.coordinate_type == ggrid.CoordinateType.CARTESIAN
+
+            assert [str(v) for v in specgrid.to_grdecl()] == values
+
+
+@pytest.mark.parametrize(
+    "inp_str, expected_unit, expected_relative",
+    [
+        ("GRIDUNIT\n 'METRES  ' '        ' /", "METRES", ggrid.GridRelative.ORIGIN),
+        ("GRIDUNIT\n METRES /", "METRES", ggrid.GridRelative.ORIGIN),
+        ("GRIDUNIT\n FEET MAP /", "FEET", ggrid.GridRelative.MAP),
+    ],
+)
+def test_gridunit(inp_str, expected_unit, expected_relative):
+    with patch("builtins.open", mock_open(read_data=inp_str)) as mock_file:
+        with open_grdecl(mock_file, keywords=["GRIDUNIT"]) as kw:
+            keyword, values = next(kw)
+            assert keyword == "GRIDUNIT"
+            gridunit = ggrid.GridUnit.from_grdecl(values)
+
+            assert gridunit.unit == expected_unit
+            assert gridunit.grid_relative == expected_relative
+
+
+@settings(deadline=None, suppress_health_check=[HealthCheck.function_scoped_fixture])
+@given(grdecl_grids())
+def test_grdecl_grid_read_write(tmp_path, grgrid):
+    tmp_file = tmp_path / "grid.grdecl"
+    grgrid.to_file(tmp_file)
+    assert ggrid.GrdeclGrid.from_file(tmp_file) == grgrid
+
+
+@given(grdecl_grids())
+def test_grdecl_grid_cylindrical_raises(grgrid):
+    assume(grgrid.specgrid.coordinate_type == ggrid.CoordinateType.CYLINDRICAL)
+    with pytest.raises(NotImplementedError, match="cylindrical"):
+        grgrid.xtgeo_coord()
+
+
+@given(
+    grdecl_grids(
+        spec=specgrids(
+            coordinates=st.just(ggrid.CoordinateType.CARTESIAN),
+            numres=st.integers(min_value=1, max_value=10),
+        )
+    )
+)
+def test_grdecl_grid_multireservoir_raises(grgrid):
+    assume(grgrid.specgrid.numres != 1)
+    with pytest.raises(NotImplementedError, match="reservoir"):
+        grgrid.xtgeo_coord()
+
+
+@given(
+    grdecl_grids(
+        spec=st.just(ggrid.SpecGrid()),
+        gunit=st.just(ggrid.GridUnit(grid_relative=ggrid.GridRelative.MAP)),
+    )
+)
+def test_grdecl_grid_maprelative_raises(grgrid):
+    assume(
+        grgrid.gridunit is not None
+        and grgrid.gridunit.grid_relative == ggrid.GridRelative.MAP
+    )
+    with pytest.raises(NotImplementedError, match="relative"):
+        grgrid.xtgeo_coord()
+
+
+@given(xtgeo_grids)
+def test_to_from_xtgeogrid_format2(xtggrid):
+    xtggrid._xtgformat2()
+    grdecl_grid = ggrid.GrdeclGrid.from_xtgeo_grid(xtggrid)
+
+    assert_allclose(grdecl_grid.xtgeo_actnum(), xtggrid._actnumsv, atol=0.02)
```

Comment:
I think the latter could be used in this case, but numpy arrays have an unusual equality operator so that `np.array([1,0]) == np.array([0,1])` can mean pointwise `or` or pointwise `and` depending on whether its enclosed in `np.and` or `np.or`. I will change this one into `assert grdecl_grid.xtgeo_actnum().tolist() == xtggrid._actnumsv` and check out how pytest.approx handles this.

Reply:
Seems like pytest.approx will force the numpy array to be serialized and the comparison to happen as lists. Changed it to use `tolist()` for actnum and `pytest.approx` for coord and zcorn.